brave/brave-core
# webpack build targets now generate a ninja depfile

This is a pre-requisite for moving some features from brave-ui to brave-core, since having to add an 'input' for each JS module used is more developer overhead than necessary.
This follows a similar pattern for chromium's webui generation, although that uses very different frameworks.

Add a new custom plugin for webpack which keeps track of every require (or import) module path, and provide a list of these for ninja. After this, build targets which use this template will not need to provide an input list of every JS / TS / CSS / etc file that is used for the build. Instead ninja will use this new generated file to keep track of which files should cause a re-build of a specific target when they are changed.

Tested with:
- Removing inputs of a webpack target
- Generated depfile contains relevant inputs
- Target only builds when an item in the generated depfile is changed

Note that the file path used in the header of the depfile must match precisely the output of the target. If this is different then the target will always be re-built as the depfile will be invalid.

Subsequently, removes 'input' param for all transpile_web_ui build targets

Fix https://github.com/brave/brave-browser/issues/3518
Closes https://github.com/brave/brave-browser/issues/2363 (removes need for it)

Reviewers: the generation is at `components/webpack/webpack-plugin-depsfile.js`

## Submitter Checklist:

- Verified that these changes build without errors on
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on

## Test Plan:
- Build
- Change a dependent webui file, e.g. components/brave_new_tab_ui/components/newTab/index.tsx (add some text to the render())
- Build again
- Verify: that the changed text is there

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.
